components/gantry/oneaxis/setup.go

```diff
@@ -0,0 +1,390 @@
+	limitHitMu      sync.Mutex
+	limitsPolled    bool
+
+	mu     sync.Mutex
+	logger golog.Logger
+	generic.Unimplemented
+}
+
+// newOneAxis creates a new one axis gantry.
+func newLimitSwitchGantry(
+	ctx context.Context,
+	cfg config.Component,
+	logger golog.Logger,
+	oneAxis *oneAxis,
+) (gantry.LocalGantry, error) {
+	conf, ok := cfg.ConvertedAttributes.(*AttrConfig)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(conf, cfg.ConvertedAttributes)
+	}
+
+	if len(conf.LimitSwitchPins) > 2 {
+		return nil, errBadNumLimitSwitches(cfg.Name, len(conf.LimitSwitchPins))
+	}
+
+	g := &limitSwitchGantry{
+		ctx:             ctx,
+		limitSwitchPins: conf.LimitSwitchPins,
+		limitHigh:       *conf.LimitPinEnabled,
+		oAx:             oneAxis,
+		logger:          logger,
+		limitsPolled:    false,
+	}
+
+	if len(g.limitSwitchPins) == 1 && (g.oAx.mmPerRevolution == 0 || g.oAx.lengthMm == 0) {
+		return nil, errDimensionsNotFound(g.oAx.name, g.oAx.mmPerRevolution, g.oAx.mmPerRevolution)
+	}
+
+	g.limitHitChecker()
+
+	return g, nil
+}
+
+func (g *limitSwitchGantry) limitHitChecker() {
+	// lock the mutex associated with this monitoring function
+	g.limitHitMu.Lock()
+	// check if the state is false
+	if !g.limitsPolled {
+		return
+	}
+	// we know we are monitoring the limit switches in this gantry,
+	// return if the limitHits are already being checked
+	g.limitsPolled = true
+	g.limitHitMu.Unlock()
+
+	var errs error
+	for {
+		select {
+		case <-g.ctx.Done():
+			return
+		default:
+		}
+		for idx := range g.limitSwitchPins {
+			hit, err := g.limitHit(g.ctx, idx)
+			if hit || err != nil {
+				g.limitHitMu.Lock()
+				g.limitsPolled = false
+				g.limitHitMu.Unlock()
+				errs := multierr.Combine(errs, err)
+				// motor operation manager cancels running
+				err = g.oAx.motor.Stop(g.ctx, nil)
+				errs = multierr.Combine(errs, err)
+				if errs != nil {
+					g.logger.Error(errs)
+					return
+				}
+			}
+		}
+	}
+}
+
+func (g *limitSwitchGantry) limitHit(ctx context.Context, offset int) (bool, error) {
+	if offset < 0 || offset > 1 {
+		return true, errBadNumLimitSwitches(g.oAx.name, offset)
+	}
+	pin, err := g.oAx.board.GPIOPinByName(g.limitSwitchPins[offset])
+	if err != nil {
+		return false, err
+	}
+	high, err := pin.Get(ctx, nil)
+
+	return high == g.limitHigh, err
+}
+
+func (g *limitSwitchGantry) testLimit(ctx context.Context, limInt int) (float64, error) {
+	defer utils.UncheckedErrorFunc(func() error {
+		return g.oAx.motor.Stop(ctx, nil)
+	})
+
+	d := -1.0 // go backwards to hit first limit switch pin
+	if limInt > 0 {
+		d *= -1 // go forwards to hit second limit switch pin
+	}
+
+	err := g.oAx.motor.GoFor(ctx, d*g.oAx.rpm, 0, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	start := time.Now()
+	for {
+		hit, err := g.limitHit(ctx, limInt)
+		if err != nil {
+			return 0, err
+		}
+		if hit {
+			err = g.oAx.motor.Stop(ctx, nil)
+			if err != nil {
+				return 0, err
+			}
+			break
+		}
+
+		elapsed := start.Sub(start)
```

Comment:
Isn't `start.Sub(start)` always going to be 0? I wonder if you meant `time.Now().Sub(start)` instead.

Reply:
good catch, thank you.